packages/world/src/systems/libraries/MoveLib.sol

```diff
@@ -151,28 +207,42 @@ library MoveLib {
     }
   }
 
-  function _gravityApplies(Vec3 playerCoord) internal view returns (bool) {
-    Vec3 belowCoord = playerCoord - vec3(0, 1, 0);
+  function runGravity(EntityId playerEntityId, Vec3 playerCoord) public {
+    if (!_gravityApplies(playerCoord)) {
+      return;
+    }
 
-    ObjectTypeId belowObjectTypeId = safeGetObjectTypeIdAt(belowCoord);
-    // Players can swim in water so we don't want to apply gravity to them
-    if (belowObjectTypeId == ObjectTypes.Water || !ObjectTypeMetadata._getCanPassThrough(belowObjectTypeId)) {
-      return false;
+    Vec3[] memory playerCoords = ObjectTypes.Player.getRelativeCoords(playerCoord);
+    EntityId[] memory playerEntityIds = _getPlayerEntityIds(playerEntityId, playerCoords);
+
+    // Remove the current player from the grid
+    for (uint256 i = 0; i < playerCoords.length; i++) {
+      ReversePlayerPosition._deleteRecord(playerCoords[i]);
     }
-    if (getPlayer(belowCoord).exists()) {
-      return false;
+
+    (Vec3 finalCoord, uint128 totalCost) = _computeGravityResult(playerCoord, 0);
+
+    uint128 currentEnergy = Energy._getEnergy(playerEntityId);
```

Comment:
We need to read the energy by calling `updatePlayerEnergy` right? Otherwise if you mine a block below a player/stacked player case, their energy will be the one thats not updated

Reply:
done, will merge!